File: ios/mindsherpa/mindsherpa/StudentProgressAPI.swift
//
//  StudentProgressAPI.swift
//  mindsherpa
//
//  Created by Claude Code on 9/17/25.
//

import Foundation
import UIKit

class StudentProgressAPI: ObservableObject {
    static let shared = StudentProgressAPI()

    @Published var isDeviceRegistered = false
    @Published var studentInfo: StudentInfo?
    @Published var isLoading = false
    @Published var lastError: String?

    private let baseURL: String
    private var deviceId: String {
        return DeviceManager.shared.deviceId
    }

    private init() {
        // Use AppConfig for consistent URL management
        self.baseURL = AppConfig.currentBaseURL
        loadStudentInfo()
    }

    // MARK: - Models

    struct StudentInfo: Codable {
        let studentId: String
        let teacherId: String
        let schoolId: String
        let firstName: String?
        let lastName: String?
        let email: String?
        let classCode: String?
        let classDetails: ClassDetails?
    }

    struct ClassDetails: Codable {
        let teacherName: String
        let teacherEmail: String
        let schoolName: String
        let programName: String
        let classCode: String
    }

    struct DeviceRegistrationRequest: Codable {
        let deviceId: String
        let platform: String
        let appVersion: String
        let deviceName: String
    }

    struct ClassJoinRequest: Codable {
        let deviceId: String
        let classCode: String
        let firstName: String
        let lastName: String
        let email: String?
    }

    struct VideoProgressRequest: Codable {
        let videoId: String
        let deviceId: String
        let watchedSeconds: Double
        let totalDuration: Double
        let isCompleted: Bool
        let courseId: String
        let lastPosition: Double
    }

    struct ApiResponse<T: Codable>: Codable {
        let success: Bool
        let data: T?
        let error: String?
        let message: String?
    }

    struct ClassJoinResponse: Codable {
        let success: Bool
        let studentId: String
        let teacherId: String
        let schoolId: String
        let message: String
        let classDetails: ClassDetails?
    }

    struct VideoProgressResponse: Codable {
        let success: Bool
        let progress: VideoProgressData
        let message: String
    }

    struct VideoProgressData: Codable {
        let videoId: String
        let deviceId: String
        let watchedSeconds: Int
        let totalDuration: Int
        let progressPercentage: Int
        let isCompleted: Bool
        let lastWatchedAt: String
        let courseId: String
    }

    // MARK: - Device Registration

    @MainActor
    func registerDevice() async {
        isLoading = true
        lastError = nil

        let request = DeviceRegistrationRequest(
            deviceId: deviceId,
            platform: "ios",
            appVersion: Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "1.0.0",
            deviceName: UIDevice.current.name
        )

        do {
            let url = URL(string: "\(baseURL)/api/progress/register-device")!
            var urlRequest = URLRequest(url: url)
            urlRequest.httpMethod = "POST"
            urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
            urlRequest.httpBody = try JSONEncoder().encode(request)

            let (data, response) = try await URLSession.shared.data(for: urlRequest)

            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                isDeviceRegistered = true
                print("✅ Device registered successfully")
            } else {
                lastError = "Failed to register device"
                print("❌ Device registration failed")
            }

        } catch {
            lastError = "Network error: \(error.localizedDescription)"
            print("❌ Device registration error: \(error)")
        }

        isLoading = false
    }

    // MARK: - Class Joining

    @MainActor
    func joinClass(classCode: String, firstName: String, lastName: String, email: String? = nil) async -> Bool {
        isLoading = true
        lastError = nil

        let request = ClassJoinRequest(
            deviceId: deviceId,
            classCode: classCode,
            firstName: firstName,
            lastName: lastName,
            email: email
        )

        do {
            let url = URL(string: "\(baseURL)/api/progress/join-class")!
            var urlRequest = URLRequest(url: url)
            urlRequest.httpMethod = "POST"
            urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
            urlRequest.httpBody = try JSONEncoder().encode(request)

            let (data, response) = try await URLSession.shared.data(for: urlRequest)

            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                let joinResponse = try JSONDecoder().decode(ClassJoinResponse.self, from: data)

                studentInfo = StudentInfo(
                    studentId: joinResponse.studentId,
                    teacherId: joinResponse.teacherId,
                    schoolId: joinResponse.schoolId,
                    firstName: firstName,
                    lastName: lastName,
                    email: email,
                    classCode: classCode,
                    classDetails: joinResponse.classDetails
                )

                saveStudentInfo()
                isLoading = false

                // Update class access in ProgressStore
                ProgressStore.shared.updateClassAccess()

                print("✅ Successfully joined class: \(classCode)")
                return true
            } else {
                let httpResponse = response as? HTTPURLResponse
                let errorData = try? JSONDecoder().decode([String: String].self, from: data)

                if httpResponse?.statusCode == 404 {
                    lastError = "This Class does not exist"
                } else {
                    lastError = errorData?["error"] ?? "Failed to join class"
                }
                print("❌ Class join failed: \(lastError ?? "Unknown error")")
            }

        } catch {
            lastError = "Network error: \(error.localizedDescription)"
            print("❌ Class join error: \(error)")
        }

        isLoading = false
        return false
    }

    // MARK: - Progress Sync

    func syncVideoProgress(videoId: String, courseId: String, watchedSeconds: Double, totalDuration: Double, isCompleted: Bool, lastPosition: Double) async {
        // Only sync if we have student info (device is linked to a student)
        guard studentInfo != nil else {
            print("📤 Skipping progress sync - no student linked")
            return
        }

        let request = VideoProgressRequest(
            videoId: videoId,
            deviceId: deviceId,
            watchedSeconds: watchedSeconds,
            totalDuration: totalDuration,
            isCompleted: isCompleted,
            courseId: courseId,
            lastPosition: lastPosition
        )

        do {
            let url = URL(string: "\(baseURL)/api/progress/video")!
            var urlRequest = URLRequest(url: url)
            urlRequest.httpMethod = "POST"
            urlRequest.setValue("application/json", forHTTPHeaderField: "Content-Type")
            urlRequest.httpBody = try JSONEncoder().encode(request)

            let (data, response) = try await URLSession.shared.data(for: urlRequest)

            if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 {
                print("📤 Video progress synced: \(videoId)")
            } else {
                print("❌ Failed to sync video progress")
            }

        } catch {
            print("❌ Progress sync error: \(error)")
        }
    }

    // MARK: - Student Info Persistence

    private func saveStudentInfo() {
        if let data = try? JSONEncoder().encode(studentInfo) {
            UserDefaults.standard.set(data, forKey: "StudentInfo")
        }
    }

    private func loadStudentInfo() {
        if let data = UserDefaults.standard.data(forKey: "StudentInfo"),
           let info = try? JSONDecoder().decode(StudentInfo.self, from: data) {
            studentInfo = info
        }
    }

    // MARK: - Helper Methods

    var isStudentLinked: Bool {
        return studentInfo != nil
    }

    var studentDisplayName: String {
        guard let info = studentInfo else { return "" }
        if let first = info.firstName, let last = info.lastName {
            return "\(first) \(last)"
        }
        return info.studentId
    }

    func clearStudentInfo() {
        studentInfo = nil
        UserDefaults.standard.removeObject(forKey: "StudentInfo")
    }
}

File: android-mindsherpa/app/src/main/java/com/skillvergence/mindsherpa/data/api/StudentProgressAPI.kt
package com.skillvergence.mindsherpa.data.api

import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.google.gson.Gson
import com.skillvergence.mindsherpa.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.util.*

/**
 * StudentProgressAPI - Handles device registration, class joining, and progress sync
 */
class StudentProgressAPI private constructor(private val context: Context) {

    companion object {
        @Volatile
        private var INSTANCE: StudentProgressAPI? = null

        fun getInstance(context: Context): StudentProgressAPI {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: StudentProgressAPI(context.applicationContext).also { INSTANCE = it }
            }
        }
    }

    private val sharedPrefs: SharedPreferences =
        context.getSharedPreferences("student_progress_api", Context.MODE_PRIVATE)
    private val gson = Gson()
    private val apiService = NetworkModule.createApiService()

    // LiveData
    private val _isDeviceRegistered = MutableLiveData<Boolean>()
    val isDeviceRegistered: LiveData<Boolean> = _isDeviceRegistered

    private val _studentInfo = MutableLiveData<StudentInfo?>()
    val studentInfo: LiveData<StudentInfo?> = _studentInfo

    private val _isLoading = MutableLiveData<Boolean>()
    val isLoading: LiveData<Boolean> = _isLoading

    private val _lastError = MutableLiveData<String?>()
    val lastError: LiveData<String?> = _lastError

    init {
        loadStudentInfo()
    }

    /**
     * Student Information Model
     */
    data class StudentInfo(
        val studentId: String,
        val teacherId: String,
        val schoolId: String,
        val firstName: String?,
        val lastName: String?,
        val email: String?,
        val classCode: String?,
        val classDetails: com.skillvergence.mindsherpa.data.api.ClassDetails? = null
    )

    /**
     * Get device ID (from DeviceManager or generate)
     */
    private fun getDeviceId(): String {
        var deviceId = sharedPrefs.getString("device_id", null)
        if (deviceId == null) {
            deviceId = "android-${UUID.randomUUID()}"
            sharedPrefs.edit().putString("device_id", deviceId).apply()
        }
        return deviceId
    }

    /**
     * Register device with backend
     */
    suspend fun registerDevice(): Boolean = withContext(Dispatchers.IO) {
        try {
            _isLoading.postValue(true)
            _lastError.postValue(null)

            val request = DeviceRegistrationRequest(
                deviceId = getDeviceId(),
                platform = "android",
                appVersion = BuildConfig.VERSION_NAME,
                deviceName = "${Build.MANUFACTURER} ${Build.MODEL}"
            )

            val response = apiService.registerDevice(request)
            if (response.isSuccessful && response.body()?.success == true) {
                _isDeviceRegistered.postValue(true)
                println("✅ Device registered successfully")
                true
            } else {
                _lastError.postValue("Failed to register device")
                println("❌ Device registration failed")
                false
            }
        } catch (e: Exception) {
            _lastError.postValue("Network error: ${e.message}")
            println("❌ Device registration error: $e")
            false
        } finally {
            _isLoading.postValue(false)
        }
    }

    /**
     * Join class with class code
     */
    suspend fun joinClass(
        classCode: String,
        firstName: String,
        lastName: String,
        email: String? = null
    ): Boolean = withContext(Dispatchers.IO) {
        try {
            _isLoading.postValue(true)
            _lastError.postValue(null)

            val request = ClassJoinRequest(
                deviceId = getDeviceId(),
                classCode = classCode.uppercase(),
                firstName = firstName,
                lastName = lastName,
                email = email
            )

            val response = apiService.joinClass(request)
            if (response.isSuccessful && response.body()?.success == true) {
                val joinResponse = response.body()!!

                val studentInfo = StudentInfo(
                    studentId = joinResponse.studentId,
                    teacherId = joinResponse.teacherId,
                    schoolId = joinResponse.schoolId,
                    firstName = firstName,
                    lastName = lastName,
                    email = email,
                    classCode = classCode.uppercase(),
                    classDetails = joinResponse.classDetails
                )

                _studentInfo.postValue(studentInfo)
                saveStudentInfo(studentInfo)

                // Update class access in ProgressStore
                com.skillvergence.mindsherpa.data.persistence.ProgressStore.getInstance(context).updateClassAccess()

                println("✅ Successfully joined class: $classCode")
                true
            } else {
                val errorBody = response.errorBody()?.string()
                val errorMessage = if (response.code() == 404) {
                    "This Class does not exist"
                } else {
                    errorBody ?: "Failed to join class"
                }
                _lastError.postValue(errorMessage)
                println("❌ Class join failed: $errorMessage")
                false
            }
        } catch (e: Exception) {
            _lastError.postValue("Network error: ${e.message}")
            println("❌ Class join error: $e")
            false
        } finally {
            _isLoading.postValue(false)
        }
    }

    /**
     * Sync video progress to backend
     */
    suspend fun syncVideoProgress(
        videoId: String,
        courseId: String,
        watchedSeconds: Double,
        totalDuration: Double,
        isCompleted: Boolean,
        lastPosition: Double
    ) = withContext(Dispatchers.IO) {
        // Only sync if we have student info (device is linked to a student)
        if (_studentInfo.value == null) {
            println("📤 Skipping progress sync - no student linked")
            return@withContext
        }

        try {
            val request = VideoProgressRequest(
                videoId = videoId,
                deviceId = getDeviceId(),
                watchedSeconds = watchedSeconds,
                totalDuration = totalDuration,
                isCompleted = isCompleted,
                courseId = courseId,
                lastPosition = lastPosition
            )

            val response = apiService.updateVideoProgress(request)
            if (response.isSuccessful) {
                println("📤 Video progress synced: $videoId")
            } else {
                println("❌ Failed to sync video progress")
            }
        } catch (e: Exception) {
            println("❌ Progress sync error: $e")
        }
    }

    /**
     * Sync podcast progress to backend
     */
    suspend fun syncPodcastProgress(
        podcastId: String,
        courseId: String,
        playbackPosition: Double,
        totalDuration: Double,
        isCompleted: Boolean
    ) = withContext(Dispatchers.IO) {
        // Only sync if we have student info
        if (_studentInfo.value == null) {
            println("📤 Skipping podcast sync - no student linked")
            return@withContext
        }

        try {
            val request = PodcastProgressRequest(
                podcastId = podcastId,
                deviceId = getDeviceId(),
                playbackPosition = playbackPosition,
                totalDuration = totalDuration,
                isCompleted = isCompleted,
                courseId = courseId
            )

            val response = apiService.updatePodcastProgress(request)
            if (response.isSuccessful) {
                println("📤 Podcast progress synced: $podcastId")
            } else {
                println("❌ Failed to sync podcast progress")
            }
        } catch (e: Exception) {
            println("❌ Podcast sync error: $e")
        }
    }

    /**
     * Student Info Persistence
     */
    private fun saveStudentInfo(studentInfo: StudentInfo) {
        val json = gson.toJson(studentInfo)
        sharedPrefs.edit().putString("student_info", json).apply()
    }

    private fun loadStudentInfo() {
        val json = sharedPrefs.getString("student_info", null)
        if (json != null) {
            try {
                val studentInfo = gson.fromJson(json, StudentInfo::class.java)
                _studentInfo.postValue(studentInfo)
            } catch (e: Exception) {
                println("❌ Failed to load student info: $e")
            }
        }
    }

    /**
     * Helper Methods
     */
    fun isStudentLinked(): Boolean {
        return _studentInfo.value != null
    }

    fun getStudentDisplayName(): String {
        val info = _studentInfo.value ?: return ""
        return if (info.firstName != null && info.lastName != null) {
            "${info.firstName} ${info.lastName}"
        } else {
            info.studentId
        }
    }

    fun clearStudentInfo() {
        _studentInfo.postValue(null)
        sharedPrefs.edit().remove("student_info").apply()
    }
}